File: clients/java-client/src/main/java/io/ehdev/conrad/client/java/VersionManagerClient.java
package io.ehdev.conrad.client.java;

import io.ehdev.conrad.client.java.exception.UnsuccessfulRequestException;
import io.ehdev.conrad.model.repo.RepoCreateModel;
import io.ehdev.conrad.model.repo.RepoResponseModel;
import org.eclipse.jgit.api.errors.GitAPIException;

import java.io.File;
import java.io.IOException;
import java.net.URISyntaxException;
import java.util.List;

public interface VersionManagerClient {

    Version findFinalVersion(File searchDir) throws IOException, GitAPIException, URISyntaxException, UnsuccessfulRequestException;

    Version findVersion(File searchDir) throws IOException, GitAPIException, URISyntaxException, UnsuccessfulRequestException;

    Version findVersion(List<String> commitIds) throws URISyntaxException, IOException, UnsuccessfulRequestException;

    Version claimVersion(List<String> commitIds, String message, String headCommitId) throws IOException, UnsuccessfulRequestException;

    Version claimVersion(File rootProjectDir) throws IOException, GitAPIException, UnsuccessfulRequestException;

    RepoResponseModel createRepository(RepoCreateModel repoCreateModel) throws IOException, UnsuccessfulRequestException;
}

File: clients/java-client/src/main/kotlin/io/ehdev/conrad/client/java/VersionManagerClient.kt
package io.ehdev.conrad.client.java

import io.ehdev.conrad.model.repository.CreateRepoRequest
import io.ehdev.conrad.model.repository.CreateRepoResponse
import org.eclipse.jgit.api.errors.GitAPIException
import java.io.File
import java.io.IOException
import java.net.URISyntaxException

interface VersionManagerClient {

    @Throws(IOException::class, GitAPIException::class, URISyntaxException::class)
    fun findFinalVersion(searchDir: File): Version

    @Throws(IOException::class, GitAPIException::class, URISyntaxException::class)
    fun findVersion(searchDir: File): Version

    @Throws(URISyntaxException::class, IOException::class)
    fun findVersion(commitIds: List<String>): Version

    @Throws(IOException::class)
    fun claimVersion(commitIds: List<String>, message: String, headCommitId: String): Version

    @Throws(IOException::class, GitAPIException::class)
    fun claimVersion(rootProjectDir: File): Version

    @Throws(IOException::class)
    fun createRepository(repoCreateModel: CreateRepoRequest): CreateRepoResponse

    @Throws(IOException::class, GitAPIException::class)
    fun tagVersion(rootProjectDir: File, version: String)
}
